### PR TITLE
Create are_models

### PR DESCRIPTION
**Pull request by alexandrosXe**
Added BILSTM and BERT ARE models

### CODE REVIEW

File: ACL2022/Models/are_models [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+      pred = Dense(1, activation = None, name = 'classification')(weighted_represantation)
+      self.model = Model(inputs=inputs, outputs=[pred, tensorflow.squeeze(att_weights)])
+      self.model.compile(loss=[self.loss, self.dummy_loss] ,
+                    optimizer=tensorflow.keras.optimizers.Adam(learning_rate=self.lr),
+                    metrics=self.metrics)
+      if self.show_summary:
+          self.model.summary()
+
+  def create_vocab(self, tokenized_texts):
+    self.vocab = list(set([w for txt in tokenized_texts for w in txt]))
```
Comment (sorensenjs) [inline ACL2022/Models/are_models]: This is not a memory efficient way of doing this, you are creating a very large list before creating a set. Use a set comprehension instead.

File: ACL2022/Models/are_models [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+    print("Vocab size: ", self.vocab_size)
+    self.w2i = {w: i+2 for i,w in enumerate(self.vocab)}
+    self.w2i[self.unk_token] = 1
+    self.w2i[self.pad_token] = 0
+    self.i2w = {i+2: self.w2i[w] for i,w in enumerate(self.vocab)}
+    self.i2w[1] = self.unk_token
+    self.i2w[0] = self.pad_token
+  
+  def to_sequences(self, tokenized_texts):
+        x = [[self.w2i[w] if w in self.w2i else self.w2i[self.unk_token] for w in t] for t in tokenized_texts]
```
Comment (sorensenjs) [inline ACL2022/Models/are_models]: this is pretty incomprehensible, maybe a comment at least.

File: ACL2022/Models/are_models [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+        x = pad_sequences(sequences=x, maxlen=self.max_seq_len, padding="post", value=0)  # padding
+        return x
+  
+  #Dummy labels are to ignore attentions
+  def fit(self, tokenized_texts, y, dummy_y, val_texts, val_y, val_dummy_y):
+    # Create vocab and lookup tables
+    self.create_vocab(tokenized_texts)
+    # turn the tokenized texts and token labels to padded sequences of indices
+    X = self.to_sequences(tokenized_texts)
+    dummy_y = pad_sequences(maxlen=self.max_seq_len, sequences=dummy_y, padding="post", value=0.0, dtype='float32')
```
Comment (sorensenjs) [inline ACL2022/Models/are_models]: mixing '' and "" quotes.

File: ACL2022/Models/are_models [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+    if  len(test_texts)%batch_size != 0:
+      texts = test_texts[counter*100:]
+      input = self.to_bert_input(texts)
+      bert_atts = tensorflow.math.reduce_mean(self.BERT(input).attentions[layer-1][:,:], axis=1)[:,0,:]
+      atts.append(bert_atts)
+    atts = [item for sublist in atts for item in sublist] #flatten
+    return atts
+
+  
+  def finetune_att_threshold(self, dev_texts, dev_position):
```
Comment (sorensenjs) [inline ACL2022/Models/are_models]: same as above - pull this out to a single function?

File: ACL2022/Models/are_models (hunk 2 above)
Comment (ipavlopoulos) [reply ACL2022/Models/are_models]: +1 @alexandrosXe please check this: https://realpython.com/python-pep8/

File: ACL2022/Models/are_models [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+      instance_toxic_char_offsets = []
+      for j,token_label in enumerate(instance):
+        if token_label == 1:
+          instance_toxic_char_offsets.extend([ch for ch in token_offsets[i][j]])
+      toxic_char_offsets.append(instance_toxic_char_offsets)
+    return toxic_char_offsets
+    
+    
+
+class BERT_ARE():
```
Comment (ipavlopoulos) [inline ACL2022/Models/are_models:218]: Some description here?

**Review by ipavlopoulos — CHANGES_REQUESTED** [review]
Please add comments at least below each method and class (incl. descriptions for the arguments and return item).

File: ACL2022/Models/are_models [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+      for i in range(self.attention_hidden_layers): #-1):
+        if i == 0:
+          x = TimeDistributed(Dense(self.rnn_layer_size, activation = None))(rnn)
+        elif i == self.attention_hidden_layers-1: #last layer
+          x = TimeDistributed(Dense(1, activation = None))(x)
+        else:
+          x = TimeDistributed(Dense(self.rnn_layer_size, activation = None))(x) #hidden layer
+      att_weights = tensorflow.nn.softmax(x, axis=1)
+      weighted_represantation = tensorflow.math.multiply(att_weights, rnn, name=None)
+      weighted_represantation = tensorflow.reduce_sum(weighted_represantation, axis = 1)
```
Comment (LeoLaugier) [inline ACL2022/Models/are_models]: represEntation

File: ACL2022/Models/are_models (hunk 2 above)
Comment (LeoLaugier) [reply ACL2022/Models/are_models]: +1

File: ACL2022/Models/are_models [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+    return history
+  
+  def predict(self, tokenized_texts):
+        predictions, atts = self.model.predict(self.to_sequences(tokenized_texts))
+        return predictions, atts
+
+  def finetune_att_threshold(self, dev_tokenized_texts, dev_token_offsets, dev_position):
+    f1s = []
+    for th in range(0,100):
+      th = th/100
```
Comment (LeoLaugier) [inline ACL2022/Models/are_models:173]: Comment?

**Review by LeoLaugier — CHANGES_REQUESTED** [review]

File: ACL2022/Models/are_models (hunk 2 above)
Comment (alexandrosXe) [reply ACL2022/Models/are_models]: added comment

File: ACL2022/Models/are_models (hunk 3 above)
Comment (alexandrosXe) [reply ACL2022/Models/are_models]: fixed

File: ACL2022/Models/are_models (hunk 4 above)
Comment (alexandrosXe) [reply ACL2022/Models/are_models]: fixed

File: ACL2022/Models/are_models (hunk 6 above)
Comment (alexandrosXe) [reply ACL2022/Models/are_models]: fixed

File: ACL2022/Models/are_models (hunk 5 above)
Comment (alexandrosXe) [reply ACL2022/Models/are_models:218]: added short description

File: ACL2022/Models/are_models (hunk 7 above)
Comment (alexandrosXe) [reply ACL2022/Models/are_models:173]: added

**Review by ipavlopoulos — APPROVED** [review]

**Pull request merged**